.buildkite/inference-benchmark-pipeline.sh

```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+set -euf
+
+# This script is executed by Buildkite to generate pipeline YAML string.
+# Document: https://buildkite.com/docs/pipelines/defining-steps#dynamic-pipelines
+#
+# Usage: CONVERT_RESULT_PATH=gs://path/to/result.tgz ./inference-benchmark-pipeline.sh
+
+# test cases to run inference
+# "<buildkite agent-type> <model filename>"
+TEST_CASES=(
+    "de10nano       lib/lib_arm.so"
+    "de10nano       lib/lib_fpga.so"
+
+    "jetson-nano    minimal_graph_with_shape.pb"
+    "jetson-tx2     minimal_graph_with_shape.pb"
+    "jetson-xavier  minimal_graph_with_shape.pb"
+
+    "raspberry-pi   lib/lib_aarch64.so"
+)
+
+# Generate setup step
+cat <<EOS
+steps:
+  - label: "inference: setup"
+    key: "setup"
+    timeout_in_minutes: "30"
+    agents:
+      - "env=benchmark"
+      - "agent-type=gcloudsdk"
+
+    artifact_paths:
+      - "convert-result.tgz"
+
+    command: |
+      gsutil -m cp "${CONVERT_RESULT_PATH}" ./
+EOS
+
+for TEST_CASE in "${TEST_CASES[@]}" ; do
+    IFS=' '
+    read AGENT MODEL <<< "${TEST_CASE}"
+
+    # Generate inference step
+    cat <<EOS
+
+
+  - label: "inference: ${AGENT} (${MODEL})"
+    depends_on: "setup"
+    timeout_in_minutes: "30"
+
+    agents:
+      - "env=benchmark"
+      - "agent-type=${AGENT}"
+
+    artifact_paths:
+      - "export/*/*/output/python/output/output.json"
+
+    command: |
+      buildkite-agent artifact download "convert-result.tgz" ./ --build $BUILDKITE_BUILD_ID
```

Comment:
```suggestion
      buildkite-agent artifact download "convert-result.tgz" ./ --build \$\${BUILDKITE_BUILD_ID}
```
I think it is better escaping `BUILDKITE_BUILD_ID` to run without `BUILDKITE_BUILD_ID` for testing this script locally.
In this situation, there are three times environment variable substitutions.

| How to write | Substitution Situation | stdout of shell | uploaded pipeline |
| --- | --- | --- | --- |
| `${BUILDKITE_BUILD_ID}` | uploading step (inference-benchmark-pipeline.sh) | `999` | `999` |
| `\${BUILDKITE_BUILD_ID}` | uploading step (buildkite-agent upload) | `${BUILDKITE_BUILD_ID}` | `999` |
| `\$\${BUILDKITE_BUILD_ID}` | inference step | `$${BUILDKITE_BUILD_ID}` | `${BUILDKITE_BUILD_ID}` |

Ref: https://buildkite.com/docs/agent/v3/cli-pipeline#environment-variable-substitution

Fortunately, `BUILDKITE BUILD_ID` is the same on each step of the same build, but I think it is better to use the env value of the inference step to avoid future potential bugs caused by referred to this script.

Reply:
Thank you so much!! I didn't think about this.